<a href="https://colab.research.google.com/github/sundarp17/Recommendation-Systems/blob/main/machine_learning/model_based_collaborative_filtering_system.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**SVD Matrix Factorization**

In [2]:
import pandas as pd
import numpy as np
import sklearn
from sklearn.decomposition import TruncatedSVD

Preparing the data

In [3]:
#u.data
columns = ['user_id','item_id','rating','timestamp']
frame  = pd.read_csv('/content/u.data',sep='\t',names=columns)
frame.head()

user_id  item_id  rating  timestamp
0      196      242       3  881250949
1      186      302       3  891717742
2       22      377       1  878887116
3      244       51       2  880606923
4      166      346       1  886397596

In [7]:
#u.item
columns = ['item_id','movie_title','release_date','video_release_date','IMDb_URL','unknown','Action','Adventure','Animation',
            'Childrens','Comedy','Crime','Documentary','Drama','Fantasy','Film_Noir','Horror','Musical','Mystery','Romance','Sci-Fi',
              'Thriller','War','Western']
movies_items = pd.read_csv('/content/u.item',sep='|',names=columns,encoding='latin-1')
movies_items.head()

item_id        movie_title release_date  ...  Thriller War  Western
0        1   Toy Story (1995)  01-Jan-1995  ...         0   0        0
1        2   GoldenEye (1995)  01-Jan-1995  ...         1   0        0
2        3  Four Rooms (1995)  01-Jan-1995  ...         1   0        0
3        4  Get Shorty (1995)  01-Jan-1995  ...         0   0        0
4        5     Copycat (1995)  01-Jan-1995  ...         1   0        0

[5 rows x 24 columns]

In [9]:
movies = movies_items[['item_id','movie_title']]
movies.head()

item_id        movie_title
0        1   Toy Story (1995)
1        2   GoldenEye (1995)
2        3  Four Rooms (1995)
3        4  Get Shorty (1995)
4        5     Copycat (1995)

In [10]:
#combining both movie datasets

combined_movies_data = pd.merge(frame,movies,on='item_id')
combined_movies_data.head()

user_id  item_id  rating  timestamp   movie_title
0      196      242       3  881250949  Kolya (1996)
1       63      242       3  875747190  Kolya (1996)
2      226      242       5  883888671  Kolya (1996)
3      154      242       3  879138235  Kolya (1996)
4      306      242       5  876503793  Kolya (1996)

In [11]:
combined_movies_data.groupby('item_id')['rating'].count().sort_values(ascending=False).head()

item_id
50     583
258    509
100    508
181    507
294    485
Name: rating, dtype: int64

In [12]:
#moviename with item id  = 50

filter = combined_movies_data['item_id']==50
combined_movies_data[filter]['movie_title'].unique()

array(['Star Wars (1977)'], dtype=object)

**Building a Utility Matrix**

In [13]:
rating_crosstab = combined_movies_data.pivot_table(values='rating',columns='movie_title',index='user_id',fill_value=0)
rating_crosstab.head()

movie_title  'Til There Was You (1997)  ...  Á köldum klaka (Cold Fever) (1994)
user_id                                 ...                                    
1                                    0  ...                                   0
2                                    0  ...                                   0
3                                    0  ...                                   0
4                                    0  ...                                   0
5                                    0  ...                                   0

[5 rows x 1664 columns]

**Transposing the matrix**

In [15]:
rating_crosstab.shape

(943, 1664)

In [16]:
x = rating_crosstab.values.T
x.shape

(1664, 943)

**Decomposing the matrix**

In [17]:
SVD = TruncatedSVD(n_components=12, random_state=17)
resultant_matrix = SVD.fit_transform(x)
resultant_matrix.shape

(1664, 12)

**Generating a corrleation matrix**

In [19]:
corr_matrix = np.corrcoef(resultant_matrix)
corr_matrix.shape

(1664, 1664)

**Isolating star wars from corrleation matrix**

In [20]:
movie_names = rating_crosstab.columns
movies_list = list(movie_names)
star_wars = movies_list.index('Star Wars (1977)')
print(star_wars)

1398


In [21]:
corr_star_wars = corr_matrix[star_wars]
corr_star_wars.shape

(1664,)

You can see that we have a vertical array of 1,664 rows, each of the rows contains a Pearson r coefficient that indicates how well each movie in the dataset correlates with Star Wars, based on user preferences.

**Recommending a highly correlated movie.**

In [22]:
list(movie_names[(corr_star_wars<1.0) & (corr_star_wars>0.9)])

['Die Hard (1988)',
 'Empire Strikes Back, The (1980)',
 'Fugitive, The (1993)',
 'Raiders of the Lost Ark (1981)',
 'Return of the Jedi (1983)',
 'Terminator 2: Judgment Day (1991)',
 'Terminator, The (1984)',
 'Toy Story (1995)']

In [23]:
list(movie_names[(corr_star_wars<1.0) & (corr_star_wars>0.95)])

['Return of the Jedi (1983)']